parser/lexer_test.go

```diff
@@ -272,24 +193,628 @@ func tokensEqual(ts1, ts2 Tokens) bool {
 	return true
 }
 
-func TestLex(t *testing.T) {
-	for _, test := range lexTests {
-		// Copy the test tokens and append an EOF token
-		testTokens := append(Tokens(nil), test.tokens...)
+func SingleTest(t *testing.T, input string, expectedError string, expected Tokens) {
+	// Copy the test tokens and append an EOF token
+	testTokens := append(Tokens(nil), expected...)
+	if len(testTokens) == 0 || testTokens[len(testTokens)-1].kind != tokenEndOfFile {
 		testTokens = append(testTokens, tEOF)
-		tokens, err := Lex(test.name, test.input)
-		var errString string
-		if err != nil {
-			errString = err.Error()
-		}
-		if errString != test.errString {
-			t.Errorf("%s: error result does not match. got\n\t%+v\nexpected\n\t%+v",
-				test.name, errString, test.errString)
-		}
-		if err == nil && !tokensEqual(tokens, testTokens) {
-			t.Errorf("%s: got\n\t%+v\nexpected\n\t%+v", test.name, tokens, testTokens)
-		}
 	}
+	tokens, err := Lex("snippet", input)
+	var errString string
+	if err != nil {
+		errString = err.Error()
+	}
+	if errString != expectedError {
+		t.Errorf("error result does not match. got\n\t%+v\nexpected\n\t%+v",
+			errString, expectedError)
+	}
+	if err == nil && !tokensEqual(tokens, testTokens) {
+		t.Errorf("got\n\t%+v\nexpected\n\t%+v", tokens, expected)
+	}
+}
+
+// TODO: test position reporting
+
+func TestEmpty(t *testing.T) {
```

Comment:
What was wrong with the old table-style tests?

Reply:
Impossible to run a single test in isolation without modifying the code.  Although Go doesn't seem to do this stuff yet, other frameworks prefer breaking stuff up as much as possible to enable individual running of tests, reporting of errors at finer granularity, running tests in parallel, and helping avoid state accumulating between tests.

Reply:
I'm ok with that.

Just FYI the way it worked before was **not** the best you could get from Table Driven Tests. All of the stuff you mentioned can be achieved with TDT. Most importantly it lacked `t.Run(tt.in, func(t *testing.T) {...` - so all the cases were treated as one test. If we added that one line with `t.Run(...` each testcase would be a subtest and we could easily run them separately and in parallel.
More info: https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks